src/main/java/org/jabref/gui/externalfiles/ImportHandler.java

```diff
@@ -303,6 +303,25 @@ private void addToGroups(List<BibEntry> entries, Collection<GroupTreeNode> group
         }
     }
 
+    public void checkCrossRefKeys(List<BibEntry> entries) {
+        for (BibEntry possibleParentEntry : entries) {
```

Comment:
As an exercise in Java programming you can eliminate the outer loop using Java Streams API. Like this: `entries.stream()...<logic to filter a parent entry>...findOne()`, and you will have an `Optional` as the result.

But, of course, eliminating one loop doesn't mean you reduced the complexity of the function, it's still $O(n^2)$ with streams

Reply:
For the mental closure: its obsolete now with the call to the citatationkeygenerator, but maybe parallelstreams would have helped with execution speed here.